File: src/org/elixir_lang/beam/chunk/Atoms.java
package org.elixir_lang.beam.chunk;

import com.intellij.openapi.util.Pair;
import org.elixir_lang.beam.chunk.atoms.Atom;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.elixir_lang.beam.chunk.Chunk.unsignedByte;
import static org.elixir_lang.beam.chunk.Chunk.unsignedInt;

public class Atoms {
    /*
     * Fields
     */

    @NotNull
    private final List<Atom> atomList;

    private Atoms(@NotNull List<Atom> atomList) {
        this.atomList = Collections.unmodifiableList(atomList);
    }

    /*
     * Static Methods
     */

    @Nullable
    public static Atoms from(@NotNull Chunk chunk, @NotNull Chunk.TypeID typeID, @NotNull Charset charset) {
        Atoms atoms = null;

        if (chunk.typeID.equals(typeID.toString()) && chunk.data.length >= 4) {
            int offset = 0;
            Pair<Long, Integer> atomCountByteCount = unsignedInt(chunk.data, offset);
            long atomCount = atomCountByteCount.first;
            offset += atomCountByteCount.second;

            List<Atom> atomList = new ArrayList<>();

            for (long i = 0; i < atomCount; i++) {
                Pair<Integer, Integer> atomLengthByteCount = unsignedByte(chunk.data[offset]);
                int atomLength = atomLengthByteCount.first;
                offset += atomLengthByteCount.second;

                String string = new String(chunk.data, offset, atomLength, charset);
                offset += atomLength;
                atomList.add(new Atom((int) i, atomLength, string));
            }

            atoms = new Atoms(atomList);
        }

        return atoms;
    }

    /*
     * Instance Methods
     */

    public int size() {
        return atomList.size();
    }

    /**
     * @param index 1-based index.  1 is reserved for {#link moduleName}
     * @return atom if it exists
     */
    @Nullable
    public Atom get(int index) {
        Atom atom = null;

        if (index >= 1) {
            // index is 1-based, so subtract 1 to get 0-based
            atom = atomList.get(index - 1);
        }

        return atom;
    }

    @Nullable
    public String moduleName() {
        String moduleName = null;

        if (atomList.size() > 0) {
            moduleName = atomList.get(0).getString();
        }

        return moduleName;
    }
}

File: src/org/elixir_lang/beam/chunk/Atoms.kt
package org.elixir_lang.beam.chunk

import org.elixir_lang.beam.chunk.Chunk.Companion.unsignedByte
import org.elixir_lang.beam.chunk.Chunk.Companion.unsignedInt
import org.elixir_lang.beam.chunk.atoms.Atom
import java.nio.charset.Charset

class Atoms private constructor(private val atomList: List<Atom>) {

    /**
     * @param index 1-based index.  1 is reserved for {#link moduleName}
     */
    operator fun get(index: Int): Atom = atomList[index - 1]

    /**
     * @param index 1-based index.  1 is reserved for {#link moduleName}
     */
    fun getOrNull(index: Int): Atom? = atomList.getOrNull(index - 1)
    fun moduleName(): String? = atomList.getOrNull(0)?.string
    fun size(): Int = atomList.size

    companion object {
        private const val TAG_U = 0

        fun from(chunk: Chunk, typeID: Chunk.TypeID, charset: Charset): Atoms? {
            var atoms: Atoms? = null

            if (chunk.typeID == typeID.toString() && chunk.data.size >= 4) {
                var offset = 0
                val atomCountByteCount = unsignedInt(chunk.data, offset)
                val atomCountSigned = atomCountByteCount.first.toInt()
                val usesTaggedLengths = (typeID == Chunk.TypeID.ATU8 && atomCountSigned < 0)
                val atomCount = if (usesTaggedLengths) -atomCountSigned else atomCountByteCount.first.toInt()
                offset += atomCountByteCount.second

                val atomList = ArrayList<Atom>()

                for (i in 1..atomCount) {
                    val atomLengthByteCount = if (usesTaggedLengths) {
                        decodeTaggedUnsignedInt(chunk.data, offset)
                    } else {
                        val lengthByteCount = unsignedByte(chunk.data[offset])
                        Pair(lengthByteCount.first, lengthByteCount.second)
                    } ?: return null

                    val atomLength = atomLengthByteCount.first
                    offset += atomLengthByteCount.second

                    val string = String(chunk.data, offset, atomLength, charset)
                    offset += atomLength
                    atomList.add(Atom(i, atomLength, string))
                }

                atoms = Atoms(atomList)
            }

            return atoms
        }

        /**
         * Decode a BEAM tagged unsigned integer (tag_u) as encoded by beam_asm:encode(?tag_u, N).
         */
        private fun decodeTaggedUnsignedInt(data: ByteArray, offset: Int): Pair<Int, Int>? {
            if (offset >= data.size) return null

            val first = data[offset].toInt() and 0xFF
            val tag = first and 0x07
            if (tag != TAG_U) return null

            return when {
                // Small: 1-byte, tag in low 3 bits, bit 3 == 0.
                (first and 0x08) == 0 -> Pair(first ushr 4, 1)

                // 2-byte: bit 3 == 1, bit 4 == 0.
                (first and 0x18) == 0x08 -> {
                    if (offset + 1 >= data.size) return null
                    val second = data[offset + 1].toInt() and 0xFF
                    val value = ((first and 0xE0) shl 3) or second
                    Pair(value, 2)
                }

                // Multi-byte (2..8 bytes) or extended length.
                else -> {
                    if ((first and 0xF8) == 0xF8) {
                        val lengthResult = decodeTaggedUnsignedInt(data, offset + 1) ?: return null
                        val lengthByteCount = lengthResult.second
                        val byteCount = lengthResult.first + 9
                        val start = offset + 1 + lengthByteCount
                        if (start + byteCount > data.size) return null
                        val value = readUnsignedBigEndian(data, start, byteCount)
                        Pair(value, 1 + lengthByteCount + byteCount)
                    } else {
                        val byteCount = (first ushr 5) + 2
                        val start = offset + 1
                        if (start + byteCount > data.size) return null
                        val value = readUnsignedBigEndian(data, start, byteCount)
                        Pair(value, 1 + byteCount)
                    }
                }
            }
        }

        private fun readUnsignedBigEndian(data: ByteArray, offset: Int, length: Int): Int {
            var value = 0
            for (i in 0 until length) {
                value = (value shl 8) or (data[offset + i].toInt() and 0xFF)
            }
            return value
        }
    }
}
